src/dftbp/dftb/hybridxc.F90

```diff
@@ -1159,34 +1165,86 @@ subroutine calculateOverlapEstimates(this, symNeighbourList, nNeighbourCamSym, i
   end subroutine calculateOverlapEstimates
 
 
+  !> Builds composite index for two nested loops (e.g. over atoms in the central cell).
+  subroutine getTwoLoopCompositeIndex(nInner, nOuter, iComposite)
+
+    !> Number of iterations in inner and outer loop
+    integer, intent(in) :: nInner, nOuter
+
+    !> Composite index for two nested loops
+    integer, intent(out), allocatable :: iComposite(:,:)
+
+    !! Auxiliary variables
+    integer :: ind, ii, jj
+
+    allocate(iComposite(2, nInner * nOuter))
+
+    ! Build up composite index for collapsing two nested loops
+    ind = 1
+    lpOuter: do ii = 1, nOuter
+      lpInner: do jj = 1, nInner
+        iComposite(1, ind) = ii
+        iComposite(2, ind) = jj
+        ind = ind + 1
+      end do lpInner
+    end do lpOuter
+
+  end subroutine getTwoLoopCompositeIndex
+
+
   !> Builds simple composite index for two nested loops over atoms in the central cell.
-  subroutine getTwoCentralCellLoopCompositeIndex(nAtom0, iAtMN)
+  subroutine getiKSiKSPrimeCompositeIndex(env, nS, nK, nKPrime, iKSComposite)
 
-    !> Number of atoms in central cell
-    integer, intent(in) :: nAtom0
+    !> Environment settings
+    type(TEnvironment), intent(in) :: env
+
+    !> Total number of spin channels
+    integer, intent(in) :: nS
+
+    !> Total number of k-points
+    integer, intent(in) :: nK, nKPrime
 
     !> Composite index for two nested loops over central cell
-    integer, intent(out), allocatable :: iAtMN(:,:)
+    integer, intent(out), allocatable :: iKSComposite(:,:)
 
-    !! Indices of atoms in central cell
-    integer :: iAtM, iAtN
+    !! Indices of spins/k-points
+    integer :: iS, iK, iKPrime
 
-    !! Auxiliary variable
-    integer :: ind
+    !! Loop counters for global and local composite index
+    integer :: indGlobal, indLocal
 
-    allocate(iAtMN(2, nAtom0**2))
+    !! Start and end index for MPI parallelization, if applicable
+    integer :: iParallelStart, iParallelEnd
 
-    ! Build up composite index iAtMN for collapsing iAtM and iAtN
-    ind = 1
-    loopM: do iAtM = 1, nAtom0
-      loopN: do iAtN = 1, nAtom0
-        iAtMN(1, ind) = iAtM
-        iAtMN(2, ind) = iAtN
-        ind = ind + 1
-      end do loopN
-    end do loopM
+  #:if WITH_MPI
```

Comment:
Better to generalize the parallel.F90 routine to return the full range in serial if not WITH_MPI

Reply:
Done.